scrape/scrape_test.go

```diff
@@ -5897,3 +5897,48 @@ func TestScrapeLoopAppendSampleLimitReplaceAllSamples(t *testing.T) {
 	}...)
 	requireEqual(t, want, resApp.resultFloats, "Appended samples not as expected:\n%s", slApp)
 }
+
+func TestScrapeLoopDisableStalenessMarkerInjection(t *testing.T) {
+	var (
+		loopDone = make(chan struct{}, 1)
+		appender = &collectResultAppender{}
+		scraper  = &testScraper{}
+		app      = func(_ context.Context) storage.Appender { return appender }
+	)
+
+	sl := newBasicScrapeLoop(t, context.Background(), scraper, app, 10*time.Millisecond)
+	// Count scrapes and terminate loop after 2 scrapes.
+	numScrapes, maxScrapes := 0, 2
+	scraper.scrapeFunc = func(ctx context.Context, w io.Writer) error {
+		numScrapes++
+		if numScrapes >= maxScrapes {
+			go sl.stop()
+			<-sl.ctx.Done()
+		}
+		if _, err := w.Write([]byte("metric_a 42\n")); err != nil {
+			return err
+		}
+		return ctx.Err()
+	}
+
+	go func() {
+		sl.run(nil)
+		loopDone <- struct{}{}
+	}()
+
+	// Disable end of run staleness markers.
+	sl.disableEndOfRunStalenessMarkers()
```

Comment:
I think this is flaky. Scheduling is non deterministic so worst case the loop runs to completion before you get to this line. I think it would be better to synchronize the test more, logic:

- start the scrape loop
- wait for some samples to appear in the appender (so we know something is being tracked)
- disable the staleness markers
- stop the scrape loop from the outside by sending a message on a channel to the scraper function

This way you don't depend on the timing at all.

Reply:
Thanks! I've applied the suggestions from your comments and the test is not flaky:
```
go test -count 100 -run "TestScrapeLoopDisableStalenessMarkerInjection" githu
b.com/prometheus/prometheus/scrape
ok      github.com/prometheus/prometheus/scrape 15.685s
```